src/Microsoft.OData.Core/ODataNotificationReader.cs

```diff
@@ -113,18 +121,44 @@ public override Task<string> ReadToEndAsync()
         /// <param name="disposing">True if called from Dispose; false if called from the finalizer.</param>
         protected override void Dispose(bool disposing)
         {
-            if (disposing)
+            if (!this.disposed && disposing)
             {
-                if (this.listener != null)
+                // Tell the listener that the stream is being disposed.
+                if (synchronous)
                 {
-                    // Tell the listener that the stream is being disposed.
-                    this.listener.StreamDisposed();
-                    this.listener = null;
+                    this.listener?.StreamDisposed();
                 }
+                else
+                {
+                    this.listener?.StreamDisposedAsync().Wait();
+                }
+
+                this.listener = null;
+                // NOTE: Do not dispose the text reader since this instance does not own it.
+                this.textReader = null;
             }
 
-            this.textReader.Dispose();
+            this.disposed = true;
             base.Dispose(disposing);

```

Comment:
Should we still call `base.Dispose(disposing)` if `this.disposed` is `true`?

Reply:
@habbes Given that the `Dispose` method is defined in a framework class and advertised as idempotent, it should be okay to call `base.Dispose(disposing)` any number of times.